sophiread/FastSophiread/src/disk_io.cpp

```diff
@@ -58,6 +58,80 @@ std::vector<char> readTPX3RawToCharVec(const std::string& tpx3file) {
   return vec;
 }
 
+/**
+ * @brief Read Timepix3 raw data from file into memory.
+ *
+ * @param tpx3file
+ * @return mapinfo_t that defines { char *, and size_t }
+ */
+mapinfo_t readTPX3RawToMapInfo(const std::string& tpx3file)
+{
+  mapinfo_t info = { -1, NULL, 0 };
+
+  // Open the file
+  std::ifstream file(tpx3file, std::ios::binary | std::ios::ate);
+
+  // Check if file is open successfully
+  if (!file.is_open()) {
+    spdlog::error("Failed to open file: {}", tpx3file);
+    exit(EXIT_FAILURE);
+  }
+
+  // Get the size of the file
+  std::streampos fileSize = file.tellg();
+  file.seekg(0, std::ios::beg);
+  spdlog::info("File size: {} bytes", static_cast<size_t>(fileSize));
+
+  // Allocate a vector to store the data
+  info.max = static_cast<size_t>(fileSize);
+  info.map = reinterpret_cast<char *>(malloc(info.max));
+
+  // Read the data from file and store it in the vector
+  if (info.map == NULL) info.max = 0;
+  else file.read(info.map, info.max);
+
+  // Close the file
+  file.close();
+
+  return info;
+}
+
+// for mmap support
+#include <fcntl.h>
+#include <sys/types.h>
+#include <sys/mman.h>
+#include <errno.h>
+#include <unistd.h>
+
+/**
+ * @brief Memory-map a Timepix3 raw data file (without pre-reading it).
+ *
+ * @param tpx3file
+ * @return mapinfo_t that defines { char *, and size_t }
+ */
+mapinfo_t mmapTPX3RawToMapInfo(const std::string& tpx3file)
+{
+  mapinfo_t info = { -1, NULL, 0 };
+
+  std::ifstream file(tpx3file, std::ios::binary);
+  if (!file.is_open()) {
+    spdlog::error("Failed to open file: {}", tpx3file);
+    exit(EXIT_FAILURE);
+  }
+
+  info.fd = open(tpx3file.c_str(),O_RDWR,0666);
+
+  if( info.fd == -1 ) { perror(tpx3file.c_str()); exit(EXIT_FAILURE); }
+  info.max = lseek(info.fd,0,SEEK_END);     // determine the sizeof the file
+  info.map = reinterpret_cast<char *>(mmap(0, info.max, PROT_READ|PROT_WRITE, MAP_SHARED, info.fd, 0));
+  // https://lemire.me/blog/2012/06/26/which-is-fastest-read-fread-ifstream-or-mmap/
+  // says to add MAP_POPULATE to make it mmap() faster...
+  // TODO: Test this
```

Comment:
Is this portion intended to support stream data from ports?

Reply:
Yes, the API of `mmap()` offers you some flags.  Minimally, we need `MAP_SHARED`.  From that reference and [this SO posting](https://stackoverflow.com/questions/58538384/using-map-populate-for-writing-to-file), adding `MAP_POPULATE` for read performance will reduce pagefaults during processing, so speeding it up.

I left the reference there to indicate that the test needs to be done still.